### PR TITLE
add built-in support for custom options, keybindings and `vim` commands

### PR DESCRIPTION
**Pull request by amtoine**
hello there :wave: :yum: 

i really love this project and wanted to contribute a bit based on a "problem" i had :blush: 

## :exclamation: related issue
i've cloned my fork of the repo, removed the header as proposed and added the tokyonight theme in `lua/custom/plugins/tokyonight.lua` :ok_hand: 

now i wanted to add
- options
- keybindings
- post-startup commands

to my config.

### without modifying the `init.lua` file as much as possible!
as you say
> ... prevent any conflicts with this init.lua if you're interested in keeping
> up-to-date with whatever is in the kickstart repo.

in my personal config, i've added
- `lua/custom/options.lua` with things like
```lua
vim.opt.relativenumber = true
```
- `lua/custom/map.lua` with things like
```lua
vim.keymap.set('n', '<c-d>', '<c-d>zz', { silent = true})
vim.keymap.set('n', '<c-u>', '<c-u>zz', { silent = true})
```
- `lua/custom/after.lua` with things like
```lua
vim.cmd([[
  " Tell Vim which characters to show for expanded TABs,
  " trailing whitespace, and end-of-lines. VERY useful!
  set listchars=tab:>\·,trail:\ ,extends:>,precedes:<,nbsp:+

  " Show problematic characters.
  set list

  " Also highlight all tabs and trailing whitespace characters.
  highlight ExtraWhitespace ctermbg=darkred guibg=darkred
  match ExtraWhitespace /\s\+$\|\t/
]])
```

## :pencil2: the changes i propose
this PR adds
- `pcall`s to the `require` function on
  - `custom/options` for `lua/custom/options.lua`
  - `custom/map` for `lua/custom/map.lua`
  - `custom/after` for `lua/custom/after.lua`
- `NOTE`s on the three `pcall`s to guide the user, with a similar message as the one for `custom/plugins/*`
- the default options of `kickstart.nvim` have been put to the top of `init.lua`, i.e. before the `lazy` setup (see the comment below for my investigation on that)

thanks to the `pcall`s, the user does not have to use these files, but can add config in them to avoid changing the main `init.lua` file!

hope you'll like it, cheers :tada: :wave:

### CODE REVIEW

**Comment by amtoine** [comment]
## :mag: a little investigation on the order of loading

**TL;DR**: the default options should come BEFORE the setup of `lazy` otherwise they overwrite the values set by the user in `after/plugin/`

i had an issue the other day about `hlsearch` not being set to `true` even though i wrote it in my config :thinking: 

the thing is
- `kickstart.nvim` sets it to `true` in `init.lua`
- but the files in `after/plugin/` are loaded BEFORE the steup of `lazy` and thus before the default options

=> the result is that the default options overwrite the user ones from `after/plugin/`...

## :wrench: some material
i've got the following patches
- [pr-files.txt](https://github.com/nvim-lua/kickstart.nvim/files/11151055/pr-files.txt) adds some files to `lua/custom/` and `after/plugin/` with a single `print("<name of file>")`
- [pr-before.txt](https://github.com/nvim-lua/kickstart.nvim/files/11151061/pr-before.txt) adds debug `print`s to `init.lua` BEFORE this PR
- [pr-after.txt](https://github.com/nvim-lua/kickstart.nvim/files/11151063/pr-after.txt) adds debug `print`s to `init.lua` AFTER this PR

## :x: before the PR
```bash
git checkout 4a37a0a  # the current master
git apply /path/to/pr-files.txt
git apply /path/to/pr-before.txt
nvim
```
gives
```
kickstart: leader key
kickstart: lazy install
kickstart: lazy setup
custom/plugins: init
All parsers are up-to-date!
after/plugin: bar
after/plugin: foo
kickstart: options
kickstart: keymap
kickstart: final setup
kickstart: done
```

### :heavy_check_mark: after the PR
```bash
git checkout feature/support-for-custom-options-bindings-and-commands
git apply /path/to/pr-files.txt
git apply /path/to/pr-after.txt
nvim
```
gives
```
kickstart: leader key
kickstart: options
custom: options
kickstart: lazy install
kickstart: lazy setup
custom/plugins: init
All parsers are up-to-date!
after/plugin: bar
after/plugin: foo
kickstart: keymap
custom: map
kickstart: final setup
custom: after
kickstart: done
```

### :exclamation: conclusions
we see that
- :x: before the PR
  - the default options of `kickstart.nvim` will overwrite the user option from `after/plugin/`
  - files in `lua/custom/*.lua` are not supported
- :heavy_check_mark: after the PR
  - files in `lua/custom/*.lua` are supported
  - files in `lua/custom/*.lua` are loaded just after their corresponding section in `init.lua`, overwriting the default with the user config
  - the options are loaded before the `after/plugin/` files

**Comment by craigmac** [comment]
after/plugin files are at the end of the `:help 'rtp`, so they are sourced last, not first.

**Comment by amtoine** [comment]
> after/plugin files are at the end of the `:help 'rtp`, so they are sourced last, not first.
>
i see that in `:help 'rtp` but this is not the behaviour i observe with `kickstart.nvim` :confused:

**Comment by amtoine** [comment]
@craigmac 

from the tip of the current `master` branch, i.e. 4a37a0a9b1d53fd355637256ac975c619226fb6d, i add the following in `after/plugin/init.lua`
```lua
print("after")
vim.o.hlsearch = true
```

then i run
```bash
nvim init.lua
```
and see `after` being printed after `All parsers are up-to-date`

but finally
```
:set hlsearch?
```
gives me `nohlsearch` and the user setting has not been used...

maybe i'm missing something obvious but i can not find where :open_mouth:

**Comment by craigmac** [comment]
start nvim and run `:scriptnames` to see the order of things sourced on startup. use `:verbose set hlsearch?` to see where it was last set (sometimes useful, but with lua it will often just say 'set from lua').

**Comment by amtoine** [comment]
`:scriptnames` gives me the following

> **Note**
> i marked the `kickstart.nvim` root with a `*` at the front

```
   1: /usr/share/nvim/runtime/ftplugin.vim
   2: /usr/share/nvim/runtime/indent.vim
   3: /etc/xdg/nvim/sysinit.vim
   4: /usr/share/nvim/archlinux.vim
*  5: ~/.local/share/git/store/github.com/amtoine/kickstart.nvim/init.lua
   6: /usr/share/nvim/runtime/filetype.lua
   7: ~/.local/share/nvim/lazy/onedark.nvim/colors/onedark.lua
   8: ~/.local/share/nvim/lazy/nvim-treesitter-textobjects/plugin/nvim-treesitter-textobjects.vim
   9: ~/.local/share/nvim/lazy/nvim-treesitter/plugin/nvim-treesitter.lua
  10: ~/.local/share/nvim/lazy/LuaSnip/plugin/luasnip.lua
  11: ~/.local/share/nvim/lazy/LuaSnip/plugin/luasnip.vim
  12: ~/.local/share/nvim/lazy/nvim-cmp/plugin/cmp.lua
  13: ~/.local/share/nvim/lazy/plenary.nvim/plugin/plenary.vim
  14: ~/.local/share/nvim/lazy/telescope.nvim/plugin/telescope.lua
  15: ~/.local/share/nvim/lazy/vim-fugitive/plugin/fugitive.vim
  16: ~/.local/share/nvim/lazy/vim-fugitive/ftdetect/fugitive.vim
  17: ~/.local/share/nvim/lazy/playground/plugin/nvim-treesitter-playground.lua
  18: ~/.local/share/nvim/lazy/Comment.nvim/plugin/Comment.lua
  19: ~/.local/share/nvim/lazy/nvim-web-devicons/plugin/nvim-web-devicons.vim
  20: ~/.local/share/nvim/lazy/neo-tree.nvim/plugin/neo-tree.vim
  21: ~/.local/share/nvim/lazy/indent-blankline.nvim/plugin/indent_blankline.vim
  22: ~/.local/share/nvim/lazy/vim-rhubarb/plugin/rhubarb.vim
  23: ~/.local/share/nvim/lazy/which-key.nvim/plugin/which-key.vim
  24: ~/.local/share/nvim/lazy/nvim-lspconfig/plugin/lspconfig.lua
  25: ~/.local/share/nvim/lazy/vim-sleuth/plugin/sleuth.vim
  26: ~/.local/share/nvim/lazy/undotree/plugin/undotree.vim
  27: /usr/share/nvim/runtime/plugin/gzip.vim
  28: /usr/share/nvim/runtime/plugin/health.vim
  29: /usr/share/nvim/runtime/plugin/man.lua
  30: /usr/share/nvim/runtime/plugin/matchit.vim
  31: /usr/share/nvim/runtime/pack/dist/opt/matchit/plugin/matchit.vim
  32: /usr/share/nvim/runtime/plugin/matchparen.vim
  33: /usr/share/nvim/runtime/plugin/netrwPlugin.vim
  34: /usr/share/nvim/runtime/plugin/rplugin.vim
  35: /usr/share/nvim/runtime/plugin/shada.vim
  36: /usr/share/nvim/runtime/plugin/spellfile.vim
  37: /usr/share/nvim/runtime/plugin/tarPlugin.vim
  38: /usr/share/nvim/runtime/plugin/tohtml.vim
  39: /usr/share/nvim/runtime/plugin/tutor.vim
  40: /usr/share/nvim/runtime/plugin/zipPlugin.vim
  41: ~/.local/share/nvim/lazy/cmp_luasnip/after/plugin/cmp_luasnip.lua
  42: ~/.local/share/nvim/lazy/cmp-nvim-lsp/after/plugin/cmp_nvim_lsp.lua
  43: ~/.local/share/nvim/lazy/tokyonight.nvim/colors/tokyonight-night.lua
  44: /usr/share/nvim/runtime/filetype.vim
  45: /usr/share/nvim/runtime/syntax/syntax.vim
  46: /usr/share/nvim/runtime/syntax/synload.vim
  47: /usr/share/nvim/runtime/autoload/provider/clipboard.vim
```

and `:verbose set hlsearch?` gives me
```
  hlsearch
        Last set from Lua
```

but still `after/plugin` appears to come before the default options of `init.lua`

**Comment by craigmac** [comment]
I don't see any `after/plugin/init.lua` being sourced in scriptnames output, so you shouldn't bee seeing that printed message... unless you mean to say you ran `nvim init.lua` when you were in the `after/plugin` directory, then you were sourcing the after/plugin/init.lua file explicitly.

Second, your runtimepath is odd, number 5.... that's an odd spot to put your `init.lua`, try putting it in an explicitly supported directory, see `:help init.lua`

```
   3: /etc/xdg/nvim/sysinit.vim
   4: /usr/share/nvim/archlinux.vim
*  5: ~/.local/share/git/store/github.com/amtoine/kickstart.nvim/init.lua
```

Also, might want to check what 3 and number 4 are doing, I don't have these on macOS, could be they are setting things up differently on your distro?

**Comment by amtoine** [comment]
> I don't see any `after/plugin/init.lua` being sourced in scriptnames output, so you shouldn't bee seeing that printed message... unless you mean to say you ran `nvim init.lua` when you were in the `after/plugin` directory, then you were sourcing the after/plugin/init.lua file explicitly.
> 
i've added back a simple `after/plugin/foo.lua` with `vim.o.hlsearch = true` in it
- no highlight when doing a search, e.g. `/vim` in `init.lua` does not highlight anything
- `:scriptnames` gives me
```
   1: /usr/share/nvim/runtime/ftplugin.vim
   2: /usr/share/nvim/runtime/indent.vim
   3: /etc/xdg/nvim/sysinit.vim
   4: /usr/share/nvim/archlinux.vim
*  5: ~/.local/share/git/store/github.com/amtoine/kickstart.nvim/init.lua
   6: /usr/share/nvim/runtime/filetype.lua
   7: ~/.local/share/nvim/lazy/onedark.nvim/colors/onedark.lua
   8: ~/.local/share/nvim/lazy/vim-fugitive/plugin/fugitive.vim
   9: ~/.local/share/nvim/lazy/vim-fugitive/ftdetect/fugitive.vim
  10: ~/.local/share/nvim/lazy/nvim-treesitter-textobjects/plugin/nvim-treesitter-textobjects.vim
  11: ~/.local/share/nvim/lazy/nvim-treesitter/plugin/nvim-treesitter.lua
  12: ~/.local/share/nvim/lazy/nvim-lspconfig/plugin/lspconfig.lua
  13: ~/.local/share/nvim/lazy/Comment.nvim/plugin/Comment.lua
  14: ~/.local/share/nvim/lazy/indent-blankline.nvim/plugin/indent_blankline.vim
  15: ~/.local/share/nvim/lazy/LuaSnip/plugin/luasnip.lua
  16: ~/.local/share/nvim/lazy/LuaSnip/plugin/luasnip.vim
  17: ~/.local/share/nvim/lazy/nvim-cmp/plugin/cmp.lua
  18: ~/.local/share/nvim/lazy/which-key.nvim/plugin/which-key.vim
  19: ~/.local/share/nvim/lazy/vim-sleuth/plugin/sleuth.vim
  20: ~/.local/share/nvim/lazy/vim-rhubarb/plugin/rhubarb.vim
  21: ~/.local/share/nvim/lazy/plenary.nvim/plugin/plenary.vim
  22: ~/.local/share/nvim/lazy/telescope.nvim/plugin/telescope.lua
  23: /usr/share/nvim/runtime/plugin/gzip.vim
  24: /usr/share/nvim/runtime/plugin/health.vim
  25: /usr/share/nvim/runtime/plugin/man.lua
  26: /usr/share/nvim/runtime/plugin/matchit.vim
  27: /usr/share/nvim/runtime/pack/dist/opt/matchit/plugin/matchit.vim
  28: /usr/share/nvim/runtime/plugin/matchparen.vim
  29: /usr/share/nvim/runtime/plugin/netrwPlugin.vim
  30: /usr/share/nvim/runtime/plugin/rplugin.vim
  31: /usr/share/nvim/runtime/plugin/shada.vim
  32: /usr/share/nvim/runtime/plugin/spellfile.vim
  33: /usr/share/nvim/runtime/plugin/tarPlugin.vim
  34: /usr/share/nvim/runtime/plugin/tohtml.vim
  35: /usr/share/nvim/runtime/plugin/tutor.vim
  36: /usr/share/nvim/runtime/plugin/zipPlugin.vim
  37: ~/.local/share/nvim/lazy/cmp_luasnip/after/plugin/cmp_luasnip.lua
  38: ~/.local/share/nvim/lazy/cmp-nvim-lsp/after/plugin/cmp_nvim_lsp.lua
* 39: ~/.local/share/git/store/github.com/amtoine/kickstart.nvim/after/plugin/foo.lua
  40: /usr/share/nvim/runtime/filetype.vim
  41: /usr/share/nvim/runtime/syntax/syntax.vim
  42: /usr/share/nvim/runtime/syntax/synload.vim
  43: /usr/share/nvim/runtime/ftplugin/lua.vim
  44: /usr/share/nvim/runtime/ftplugin/lua.lua
  45: /usr/share/nvim/runtime/indent/lua.vim
  46: /usr/share/nvim/runtime/autoload/provider/clipboard.vim
```

> Second, your runtimepath is odd, number 5.... that's an odd spot to put your `init.lua`, try putting it in an explicitly supported directory, see `:help init.lua`
> 
this is my `git` repos local store, i have the `kickstart.nvim` symlinked to `~/.config/nvim/` with
```bash
ln -s $(pwd) ~/.config/nvim
```
from inside the root of `kickstart.nvim`

could the symlink give strange behaviour? :open_mouth: 

> ```
>    3: /etc/xdg/nvim/sysinit.vim
>    4: /usr/share/nvim/archlinux.vim
> *  5: ~/.local/share/git/store/github.com/amtoine/kickstart.nvim/init.lua
> ```
> 
> Also, might want to check what 3 and number 4 are doing, I don't have these on macOS, could be they are setting things up differently on your distro?
>
i have no idea to be honest :thinking: 
i'm on arch linux and never heard of that cool command you gave until yesterday :yum:

**Comment by amtoine** [comment]
same thing from `~/.config/nvim` without any symlink
```
   1: /usr/share/nvim/runtime/ftplugin.vim
   2: /usr/share/nvim/runtime/indent.vim
   3: /etc/xdg/nvim/sysinit.vim
   4: /usr/share/nvim/archlinux.vim
*  5: ~/.config/nvim/init.lua
   6: /usr/share/nvim/runtime/filetype.lua
   7: ~/.local/share/nvim/lazy/onedark.nvim/colors/onedark.lua
   8: ~/.local/share/nvim/lazy/vim-fugitive/plugin/fugitive.vim
   9: ~/.local/share/nvim/lazy/vim-fugitive/ftdetect/fugitive.vim
  10: ~/.local/share/nvim/lazy/LuaSnip/plugin/luasnip.lua
  11: ~/.local/share/nvim/lazy/LuaSnip/plugin/luasnip.vim
  12: ~/.local/share/nvim/lazy/nvim-cmp/plugin/cmp.lua
  13: ~/.local/share/nvim/lazy/which-key.nvim/plugin/which-key.vim
  14: ~/.local/share/nvim/lazy/Comment.nvim/plugin/Comment.lua
  15: ~/.local/share/nvim/lazy/plenary.nvim/plugin/plenary.vim
  16: ~/.local/share/nvim/lazy/telescope.nvim/plugin/telescope.lua
  17: ~/.local/share/nvim/lazy/indent-blankline.nvim/plugin/indent_blankline.vim
  18: ~/.local/share/nvim/lazy/nvim-treesitter-textobjects/plugin/nvim-treesitter-textobjects.vim
  19: ~/.local/share/nvim/lazy/nvim-treesitter/plugin/nvim-treesitter.lua
  20: ~/.local/share/nvim/lazy/nvim-lspconfig/plugin/lspconfig.lua
  21: ~/.local/share/nvim/lazy/vim-rhubarb/plugin/rhubarb.vim
  22: ~/.local/share/nvim/lazy/vim-sleuth/plugin/sleuth.vim
  23: /usr/share/nvim/runtime/plugin/gzip.vim
  24: /usr/share/nvim/runtime/plugin/health.vim
  25: /usr/share/nvim/runtime/plugin/man.lua
  26: /usr/share/nvim/runtime/plugin/matchit.vim
  27: /usr/share/nvim/runtime/pack/dist/opt/matchit/plugin/matchit.vim
  28: /usr/share/nvim/runtime/plugin/matchparen.vim
  29: /usr/share/nvim/runtime/plugin/netrwPlugin.vim
  30: /usr/share/nvim/runtime/plugin/rplugin.vim
  31: /usr/share/nvim/runtime/plugin/shada.vim
  32: /usr/share/nvim/runtime/plugin/spellfile.vim
  33: /usr/share/nvim/runtime/plugin/tarPlugin.vim
  34: /usr/share/nvim/runtime/plugin/tohtml.vim
  35: /usr/share/nvim/runtime/plugin/tutor.vim
  36: /usr/share/nvim/runtime/plugin/zipPlugin.vim
  37: ~/.local/share/nvim/lazy/cmp_luasnip/after/plugin/cmp_luasnip.lua
  38: ~/.local/share/nvim/lazy/cmp-nvim-lsp/after/plugin/cmp_nvim_lsp.lua
* 39: ~/.config/nvim/after/plugin/foo.lua
  40: /usr/share/nvim/runtime/filetype.vim
  41: /usr/share/nvim/runtime/syntax/syntax.vim
  42: /usr/share/nvim/runtime/syntax/synload.vim
  43: /usr/share/nvim/runtime/ftplugin/lua.vim
  44: /usr/share/nvim/runtime/ftplugin/lua.lua
  45: /usr/share/nvim/runtime/indent/lua.vim
  46: /usr/share/nvim/runtime/autoload/provider/clipboard.vim
```
but still no highlight

**Comment by amtoine** [comment]
hope i do not steal your time here...

i did not open an issue because i had done the same experiments and had the conclusion that the order of the options should be changed in `init.lua` :confused:

**Comment by feoh** [comment]
I think what they're saying is that they used the after/plugins mechanism but weren't seeing the highlighting behavior they're looking for.

Let me take a swing at digging into this :)

**Comment by feoh** [comment]
@tjdevries I don't know that the proposed solution is the right way to go, but putting vim.o.hlsearch = true in after/plugins/default.lua doesn't work.

I see in runtimepath that $XDG_CONFIG/nvim/init.lua sets hlsearch = false, but I'd THINK that toggling the option in after/plugins would have the desired effect.

I'll keep digging as I have time as I'd like to understand why the above behavior occurs.

**Comment by amtoine** [comment]
@tjdevries @feoh 
thanks a lot for trying to dig into this :relieved: 

> I don't know that the proposed solution is the right way to go,
>
i'd be happy to change my solution, of course :yum: 

> putting vim.o.hlsearch = true in after/plugins/default.lua doesn't work.
>
that is exactly what i'm seing!!

> I'll keep digging as I have time as I'd like to understand why the above behavior occurs.
>
:sparkles:

**Comment by tjdevries** [comment]
no, not `after/plugins/default.lua` but `after/plugin/default.lua` I think you have the wrong path?

**Comment by amtoine** [comment]
@tjdevries 
from the current `master` branch, i.e. on 36842aa368cbe3df3059c25db1bd2d15bc731a94, i've applied
```diff
diff --git a/after/plugin/default.lua b/after/plugin/default.lua
new file mode 100644
index 0000000..7a3c801
--- /dev/null
+++ b/after/plugin/default.lua
@@ -0,0 +1 @@
+vim.o.hlsearch = true
```
but still the same behaviour, i.e. if i `/nvim` in `init.lua`, i only see the first match but i soon as i run `:set hlsearch`, they all appear :open_mouth: 

same with `after/plugins/default.lua` :thinking:

**Comment by feoh** [comment]
I agree that this is a serious problem a number of us are seeing but I do not think the pcall strategy is the way to go.

From my simplistic viewpoint, it's like trying to use velcro to attach things that should actually be joined with nuts and bolts.

I'm going to close this PR. Let's keep working this in #300.
